core/src/main/java/ai/timefold/solver/core/impl/domain/variable/cascade/CascadeUpdateVariableListener.java

```diff
@@ -33,12 +35,25 @@ public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object
     public void afterVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
         var currentEntity = entity;
         while (currentEntity != null) {
-            scoreDirector.beforeVariableChanged(currentEntity, sourceShadowVariableDescriptor.getVariableName());
-            Object oldValue = sourceShadowVariableDescriptor.getValue(currentEntity);
+            List<Object> oldValueList = new ArrayList<>(sourceShadowVariableDescriptorList.size());
+            for (ShadowVariableDescriptor<Solution_> sourceShadowVariableDescriptor : sourceShadowVariableDescriptorList) {
+                scoreDirector.beforeVariableChanged(currentEntity, sourceShadowVariableDescriptor.getVariableName());
+                oldValueList.add(sourceShadowVariableDescriptor.getValue(currentEntity));
+            }
             listenerUpdateMethod.executeGetter(currentEntity);
-            Object newValue = sourceShadowVariableDescriptor.getValue(currentEntity);
-            scoreDirector.afterVariableChanged(currentEntity, sourceShadowVariableDescriptor.getVariableName());
-            if (Objects.equals(oldValue, newValue)) {
+            List<Object> newValueList = new ArrayList<>(sourceShadowVariableDescriptorList.size());
+            for (ShadowVariableDescriptor<Solution_> sourceShadowVariableDescriptor : sourceShadowVariableDescriptorList) {
+                scoreDirector.afterVariableChanged(currentEntity, sourceShadowVariableDescriptor.getVariableName());
+                newValueList.add(sourceShadowVariableDescriptor.getValue(currentEntity));
+            }
+            boolean hasChange = false;
+            for (int i = 0; i < oldValueList.size(); i++) {
+                if (!Objects.equals(oldValueList.get(i), newValueList.get(i))) {
+                    hasChange = true;
+                    break;
+                }
+            }
```

Comment:
Arguably you can make this check when you're reading the new value, completely avoiding this loop and improving performance.